File: python/3294_convert_doubly_linked_list_to_array_ii.py
# Problem 3294: Convert Doubly Linked List to Array II
# https://leetcode.com/problems/convert-doubly-linked-list-to-array-ii/

from typing import List, Optional

# Definition for a Node.
class Node:
    def __init__(self, val, prev=None, next=None):
        self.val = val
        self.prev = prev
        self.next = next

class Solution:
    def to_array(self, node: Optional[Node]) -> List[int]:
        # 先宣告一個空 list
        result = []
        
        # 選取往前的 node
        prev_node = node
        # 跑迴圈, 如果確實有前面的 node 就把值加到最前面, 並往前移一個
        while prev_node:
            result.insert(0, prev_node.val)
            prev_node = prev_node.prev
        
        # 選取往後的 node
        next_node = node.next
        # 跑迴圈, 如果確實有後面的 node 就把值加到最後面, 並往後移一個
        while next_node:
            result.append(next_node.val)
            next_node = next_node.next

        return result

File: c++/3294_ConvertDoublyLinkedListToArrayII.cpp
// Problem 3294: Convert Doubly Linked List to Array II
// https://leetcode.com/problems/convert-doubly-linked-list-to-array-ii/

#include <string>
#include <vector>
using namespace std;

// Definition for doubly-linked list.
class Node
{
public:
    int val;
    Node *prev;
    Node *next;
    Node() : val(0), next(nullptr), prev(nullptr) {}
    Node(int x) : val(x), next(nullptr), prev(nullptr) {}
    Node(int x, Node *prev, Node *next) : val(x), next(next), prev(prev) {}
};

class Solution
{
public:
    vector<int> toArray(Node *node)
    {
        vector<int> result;

        Node *prev_node = node;
        while (prev_node)
        {
            result.insert(result.begin(), prev_node->val);
            prev_node = prev_node->prev;
        }

        Node *next_node = node->next;
        while (next_node)
        {
            result.push_back(next_node->val);
            next_node = next_node->next;
        }

        return result;
    }
};